plutus-core-spec/plutus-core-specification.tex

```diff
@@ -1033,6 +1025,81 @@ \section{Examples}
 % to \verb|ifix|.
 
+\section{Untyped Plutus Core}
+\label{sec:untyped-plc}
+For code transmission and on-chain execution it is important to have
+small programs.  In this section we define an untyped version of
+Plutus Core together with a type erasure function converting typed
+Plutus Core terms to untyped ones.
+
+As mentioned in Section~\ref{sec:blockchain-issues}, it is important to
+limit the consumption of computational resources in on-chain code.  A
+future version of this document will include a cost model which will
+provide fine-grained costs for individual operations and built-in
+functions, enabling accurate (dynamic) monitoring of execution costs.
+
+\subfile{figures/UntypedGrammar}
+
+\bigskip
+\noindent The grammar is essentially that of typed Plutus Core with all types
+and type-related operations discarded.  However, we require two new
+items: \texttt{delay} and \texttt{force}.  These are needed because
+type-level abstraction (\texttt{abs}) and application (\texttt{\{\}})
+have a computational effect.  The body of a type abstraction will not
+be executed until the type is instantiated; thus we cannot just
+replace $\abs{\alpha}{K}{M}$ with $M$, since the latter is evaluated
+immediately.  As we will see below, \texttt{delay} and \texttt{force}
+are used to preserve the effects of type abstraction and
+instantiation (a more standard strategy of abstracting over the unit value was
+rejected because it is less space-efficient).%
+%% \footnote{A more standard, and semantically equivalent,
+%% way to implement \texttt{delay} and \texttt{force} would be to use
+%% $\lamU{u}{e}$ instead of $\delay{e}$ and $\appU{e}{\lamU{x}{x}}$
+%% instead of $\force{e}$.  This has the
+%% advantage of not requiring any new additions to the language, but is
+%% less space-efficient.  Compared with erasing \texttt{abs}
+%% and \texttt{\{\}} completely (which is semantically unsound), using
+%% \texttt{delay} and \texttt{force} increasese script size by about 10\%, but
+%% abstacting over the unit value increases sizes by about 20\%.  Small
+%% script sizes are important, so we opted for the former.}
+
+As with typed Plutus Core, we can show that values (terms which can
+not be reduced) have a particular syntactic form: see
+Figure~\ref{fig:untyped-values}.
+
+\subsection{Type erasure}
+Figure~\ref{fig:type-erasure} defines a type erasure operation which
+converts a typed Plutus Core term $M$ into an untyped one
+$\erase{M}$ (objects from the typed grammar on the left of $=$,
+untyped on the right).  %\footnotemark[4].
+We extend this to sequences of
+terms in the obvious way: given $M^* = M_1 \ldots M_n$ we define
+$\erase{M^*} = \erase{M_1} \ldots \erase{M_n}$.  For a typed program
+$P$ we use $\erase{P}$ to denote the untyped program with the same
+version whose body is the erasure of the body of $P$. The erasure operation
+preserves semantics in an appropriate sense.  It can be shown that if $P$
```

Comment:
We haven't actually shown this, but I'm confident in the abilities of our formalisers.